### PR TITLE
Fix wasm backend invokes

### PR DESCRIPTION
**Pull request by jgravelle-google**

### CODE REVIEW

File: emscripten.py [hunk 1]
```diff
@@ -1707,16 +1709,21 @@ def math_fix(g):
...
 STACK_MAX = STACK_BASE;
 HEAP32[%d >> 2] = STACKTOP;
-Runtime.stackAlloc = asm['stackAlloc'];
-Runtime.stackSave = asm['stackSave'];
-Runtime.stackRestore = asm['stackRestore'];
-Runtime.establishStackSpace = asm['establishStackSpace'];
 ''' % shared.Settings.GLOBAL_BASE)
-
+  if settings['BINARYEN_ASYNC_COMPILATION']:
+    funcs_js.append("__ATINIT__.unshift(function() {\n"
```
Comment (dschuff) [inline emscripten.py]: Is there an equivalent of this for asm2wasm? Seems like it should work the same either way?

Comment (jgravelle-google) [inline emscripten.py]: There is: https://github.com/jgravelle-google/emscripten/blob/1d81afeb464f9247698eff4780f0f3ef7781c164/emscripten.py#L1326

It uses `Module` instead of `asm`, and we just get done setting `Module["setTempRet0"]` to be a function that proxies a call to `Module["asm"]["setTempRet0"]`. `asm`/`Module["asm"]` hasn't been set yet for async compilation, but when thunking it there's no problems?

Trying that for wasm backend causes regressions from this PR for reasons I don't fully understand (specifically not having a `setTempRet0` on the global asm object).
... Ok digging deeper into that, this fixes library.js functions that with asmjs get included in the module, such as `saveSetjmp`. The specific part here is `asm = Module['asm']` at init, which should probably go in receiveInstance. So that would be a problem for asm2wasm if the `asm` references in library asm code didn't get sucked in to the wasm code.

File: src/preamble.js [hunk 2]
```diff
@@ -2223,6 +2223,7 @@ function integrateWasmJS(Module) {
       exports = instance.exports;
       if (exports.memory) mergeMemory(exports.memory);
       Module['asm'] = exports;
+      asm = exports;
```
Comment (kripken) [inline src/preamble.js]: why is this needed?

Comment (jgravelle-google) [inline src/preamble.js]: Because without it functions in js libraries that are marked `__asm` and use `setTempRet` (https://github.com/kripken/emscripten/blob/6dc4ac5f9e4d8484e273e4dcc554f809738cedd6/src/parseTools.js#L1296) look for it in the wrong place when we use async compilation, because we set global-level `asm` to the results of `doNativeWasm` which is empty, but we don't fix it up afterward.

That should probably be guarded by `#if WASM_BACKEND`.

Comment (kripken) [inline src/preamble.js]: How about changing that code in parseTools from looking at `asm` to looking at `Module['asm']`? That should be the proper way to look, allowing for updates.

Comment (dschuff) [inline src/preamble.js]: +1, in general I think we should try to avoid having our code use the global variables as much as possible.

Comment (jgravelle-google) [inline src/preamble.js]: Sounds good, done in https://github.com/kripken/emscripten/pull/5099/commits/fdcedf47f0d303727d6727eee0ac136b3a7e7d9e

Ideally we could scope the global `asm` variable to be local to instantiation. Alternatively we can avoid declaring `var asm` at all, and just use `Module['asm']` directly.
##
... ok tried that in 50bed1839513b5627beba9d7c6e6b93d6a87123f which seems to work, what do people think?
Reasons I like it:
* removes the ability to get use `asm` wrongly, forcing use of `Module['asm']`

Reasons we should probably revert that commit:
* it's more different from asm.js (if we want to do this we can do so for asm.js as well)
* any non-Emscripten js libraries that used the old `asm` will break

Comment (kripken) [inline src/preamble.js]: I think it's a good idea to remove the `var asm`, but yeah, let's not create a difference between the wasm backend and asm.js. We should do it in both (probably as a separate PR later I guess).

Comment (jgravelle-google) [inline src/preamble.js]: Sounds good, reverted in https://github.com/kripken/emscripten/pull/5099/commits/63f2f813b313d6e6303a3c40891c4155312a499b

**Comment by kripken** [comment]
Cool, lgtm.
